apps/web/app/.server/service/proposals.ts

```diff
@@ -67,3 +76,111 @@ export async function nowInSeconds() {
   const block = await l1Rpc.getBlock({ blockTag: "latest" });
   return block.timestamp;
 }
+
+export type ProposalDataResponse = {
+  l2ProposalId: Hex;
+} & (
+  | {
+      ok: true;
+      data: StartUpgradeData;
+      l1ProposalId: Hex;
+      error: null;
+    }
+  | {
+      ok: false;
+      error: string;
+      data: null;
+      l1ProposalId: null;
+    }
+);
+
+async function extractProposalData(txHash: Hex, l2ProposalId: Hex): Promise<ProposalDataResponse> {
+  console.log("txHash", txHash);
+  const receipt = await l2Rpc.getTransactionReceipt({ hash: txHash });
+  const logProof = await fetchLogProof(txHash, 0);
+
+  const l1MessageEventId = "0x3a36e47291f4201faf137fab081d92295bce2d53be2c6ca68ba82c7faa9ce241";
+  const bodyLog = receipt.logs.find((l) => l.topics[0] === l1MessageEventId);
+  if (!bodyLog) {
+    return {
+      l2ProposalId: l2ProposalId,
+      l1ProposalId: null,
+      ok: false,
+      error: `No message sent to l1 found for tx ${txHash}`,
+      data: null,
+    };
+  }
+
+  if (!logProof) {
+    return {
+      l2ProposalId: l2ProposalId,
+      l1ProposalId: null,
+      ok: false,
+      error: `log proof was not found for tx ${txHash}`,
+      data: null,
+    };
+  }
+
+  if (receipt.l1BatchNumber === null) {
+    return {
+      l2ProposalId: l2ProposalId,
+      l1ProposalId: null,
+      ok: false,
+      error: "missing batch number",
+      data: null,
+    };
+  }
+
+  if (receipt.l1BatchTxIndex === null) {
+    console.warn(`Missing l1BatchTxIndex for tx ${txHash}`);
```

Comment:
We have a proper logger function that can be imported for this case.

Reply:
:+1: